twister: fixes to statistics and counting, reduce platform scan verbosity

nashif opened the pull request:
- **twister: count filtered testcases as such, not as skipped**
- **twister: use consistent language for test configurations**
- **twister: do not report filtered cases in summary**
- **twister: stats: not run tests are also part not executed summary**
- **twister: remove verbose debug message about adding platforms**


Fixes #81128

nashif committed:
twister: remove verbose debug message about adding platforms

This is very verbose and very long sometimes getting the way when trying
to debug a problem.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: count filtered testcases as such, not as skipped

When a suite is filtered, its testcases are also filtered and not
skipped.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: use consistent language for test configurations

In the summary, use "configurations" instead of instances.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: do not report filtered cases in summary

It is confusing to report filtered testcases as  testcases that were
selected but not exexuted. If they are filtered, then there should not
be considered as selected.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: stats: not run tests are also part not executed summary

Add notrun tests to the count of non-executed tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: status inconsistencies are now warnings

Do not report status issues as errors, very confusing and developer end
up looking at the wrong thing, instead, treat those as warnings and
count them and report them at the end.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

golowanow commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -615,12 +615,11 @@ def summary(self, results, ignore_unrecognized_sections, duration):
+            if results.skipped_cases or results.notrun_cases:
                 logger.info(
-                    f'{results.skipped_cases + results.filtered_cases} selected test cases not executed:' \
+                    f'{results.skipped_cases + results.notrun_cases} selected test cases not executed:' \
```
with `./scripts/twister -p sam_e70_xplained/same70q21 -T tests/subsys/dsp/basicmath`
this PR:
```
Total test cases: 949
├─ Filtered test cases:   1
└─ Executed test cases: 948
   ├─ Passed test cases:         0
   ├─ Built only test cases:   948
INFO    - 948 selected test cases not executed: 948 not run (built only).
```
'selected not executed` built-only under its 'Executed` root looks strange, mb it should be
`└─ Selected test cases: 948` ?

before it was also strange, but differently )
```
INFO    - 1 selected test cases not executed: 1 filtered, 948 not run (built only).
```

nashif replied:
not sure what the problem is? This all looks correct.

nashif replied:
oh, you mean this:
```
└─ Executed test cases: 948
   ├─ Passed test cases:         0
   ├─ Built only test cases:   948
```

nashif replied:
yeah, this looks better.

nashif committed:
twister: stats: 'Executed test cases' -> 'Completed test cases'

Built tests are not executed, change column title.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twisters: tests: adapt twister testsuite

Adapt tests for new changes in twister.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: stats: remove double counting of statuses

We have been double counting some statuses, remove to get the stats
right.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

golowanow commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -615,12 +615,11 @@ def summary(self, results, ignore_unrecognized_sections, duration):
                 logger.info(
-                    f'{results.skipped_cases + results.filtered_cases} selected test cases not executed:' \
+                    f'{results.skipped_cases + results.notrun_cases} selected test cases not executed:' \
                     f'{" " + str(results.skipped_cases) + " skipped" if results.skipped_cases else ""}' \
```
there is one confusion left in total info with 'skipped' printed here as 'selected':
```
├─ Skipped test cases:   10                   
└─ Selected test cases: 130
...
INFO    - 10 selected test cases not executed: 10 skipped.
```

nashif replied:
why is there confusion? A selected test can run and be skipped.

nashif replied:
I guess I know what you mean now, made the change

golowanow reviewed (commented):
Please remove 'Trivial' label - it is not so anymore for most of 9 commits here.

nashif committed:
twister: stats: skipped testcases are part of selected group

Count skipped cases as part of selected group in stats.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

golowanow commented:
just moving 'skipped' under 'selected' disbalancing the totals, eg.
```
Total test cases: 276
├─ Filtered test cases: 136
└─ Selected test cases: 130
   ├─ Passed test cases:       124
   ├─ Skipped test cases:       10
   ├─ Built only test cases:     0
   ├─ Blocked test cases:        0
   ├─ Failed test cases:         0
   ├─ Errors in test cases:      0
   ├──── The following test case statuses should not appear in a proper execution ───
   └─ Statusless test cases:     6
```
total 276 != 266, and  selected 130 != 140

nashif committed:
twister: stats: skipped_configs -> filtered_configs

Rename variables leading to confusion between skipped and filtered
suites.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: stats: more renames skipped -> filtered

Additional variable renames leading to confusion between skipped and
filtered.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: stats: fix suite statistics

suite stats were not correct, a mixup between skipped and filtered
suites was leading to inconsistent numbers. This is now fixed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
-        print(f"└─ {'Completed test suites: ':<37}{completed_configs:>{processed_suites_n_length}}")
-        print(f"   ├─ {'Filtered test suites (at runtime): ':<37}{self.skipped_runtime:>{completed_suites_n_length}}")
+        print(f"└─{'Filtered test suites: ':<21}{self.filtered_configs}")
+        print(f"  ├─ {'Filtered test suites (static): ':<37}{self.filtered_static:>{filtered_suites_n_length}}")
+        print(f"  └─ {'Filtered test suites (at runtime): ':<37}{self.filtered_runtime:>{filtered_suites_n_length}}")
```
Add `│` before the first space, so the tree is intact.

```python
print(f"│  ├─ {'Filtered test suites (static): ':<37}{self.filtered_static:>{filtered_suites_n_length}}")
print(f"│  └─ {'Filtered test suites (at runtime): ':<37}{self.filtered_runtime:>{filtered_suites_n_length}}")
```

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
-        print(f"├─ {'Filtered test suites (static): ':<37}{self.skipped_filter:>{processed_suites_n_length}}")
-        print(f"└─ {'Completed test suites: ':<37}{completed_configs:>{processed_suites_n_length}}")
-        print(f"   ├─ {'Filtered test suites (at runtime): ':<37}{self.skipped_runtime:>{completed_suites_n_length}}")
+        print(f"└─{'Filtered test suites: ':<21}{self.filtered_configs}")
```
Add space after `└─`. Change `└` into `├`. Change `21` into a number bigger than the previous tree level (`21`), e.g. `23`. Right-align the numbers so that the same tree level numbers can be easily found.
`print(f"├─ {'Filtered test suites: ':<23}{self.filtered_configs:>{suites_n_length}}")`

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
+        print(f"└─{'Filtered test suites: ':<21}{self.filtered_configs}")
+        print(f"  ├─ {'Filtered test suites (static): ':<37}{self.filtered_static:>{filtered_suites_n_length}}")
+        print(f"  └─ {'Filtered test suites (at runtime): ':<37}{self.filtered_runtime:>{filtered_suites_n_length}}")
+        print(f"└─ {'Selected test suites: ':<37}{completed_configs:>{completed_suites_n_length}}")
```
Change `37` into a number that's the same as the previous column width of the same tree level - `21` before further changes, `23` after them.
`print(f"└─ {'Selected test suites: ':<23}{completed_configs:>{completed_suites_n_length}}")`

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
-        executed_cases = self.cases - self.skipped_cases - self.filtered_cases
-        completed_configs = self.done - self.skipped_filter
+        selected_cases = self.cases - self.filtered_cases
+        completed_configs = self.done - self.filtered_static - self.filtered_runtime
```
Should be changed to `selected_configs`.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
         # Find alignment length for aesthetic printing
         suites_n_length = self._find_number_length(self.total if self.total > self.done else self.done)
-        processed_suites_n_length = self._find_number_length(self.done)
         completed_suites_n_length = self._find_number_length(completed_configs)
```
Should be changed to `selected_suites_n_length`.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -143,49 +145,60 @@ def _find_number_length(n):
-        print(f"   {'├' if self.none_cases or self.started_cases else '└'}─ {'Errors in test cases: ':<25}{self.error_cases:>{executed_cases_n_length}}")
+        print(f"└─ {'Selected test cases: ':<21}{selected_cases:>{total_cases_n_length}}")
+        print(f"   ├─ {'Passed test cases: ':<25}{self.passed_cases:>{selected_cases_n_length}}")
+        print(f"   ├─ {'Skipped test cases: ':<25}{self.skipped_cases:>{total_cases_n_length}}")
```
Should use `selected_cases_n_length` instead of `total_cases_n_length`, as the rest of its tree level.

LukaszMrugala reviewed (dismissed):
Looks like it's working, so I only have some aesthetical fixes.

Before and after the summary changes:

```
--------------------------------------------------
Total test suites:     27
Processed test suites: 27
└─Filtered test suites: 11
  ├─ Filtered test suites (static):       10
  └─ Filtered test suites (at runtime):    1
└─ Selected test suites:                16
   ├─ Skipped test suites:                  0
   ├─ Passed test suites:                  13
   ├─ Built only test suites:               1
   ├─ Failed test suites:                   1
   └─ Errors in test suites:                1
----------------------      ----------------------
Total test cases: 176
├─ Filtered test cases:  29
└─ Selected test cases: 147
   ├─ Passed test cases:       119
   ├─ Skipped test cases:        1
   ├─ Built only test cases:    16
   ├─ Blocked test cases:        4
   ├─ Failed test cases:         2
   ├─ Errors in test cases:      0
   ├──── The following test case statuses should not appear in a proper execution ───
   └─ Statusless test cases:     5
--------------------------------------------------
```


```
--------------------------------------------------
Total test suites:     27
Processed test suites: 27
├─ Filtered test suites:    11
│  ├─ Filtered test suites (static):       10
│  └─ Filtered test suites (at runtime):    1
└─ Selected test suites:    16
   ├─ Skipped test suites:                  0
   ├─ Passed test suites:                  13
   ├─ Built only test suites:               1
   ├─ Failed test suites:                   1
   └─ Errors in test suites:                1
----------------------      ----------------------
Total test cases: 176
├─ Filtered test cases:  29
└─ Selected test cases: 147
   ├─ Passed test cases:       119
   ├─ Skipped test cases:        1
   ├─ Built only test cases:    16
   ├─ Blocked test cases:        4
   ├─ Failed test cases:         2
   ├─ Errors in test cases:      0
   ├──── The following test case statuses should not appear in a proper execution ───
   └─ Statusless test cases:     5
--------------------------------------------------
```

nashif committed:
twister: stats: use anytree to create summary

Do not create the tree structure manually, use anytree instead.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> Looks like it's working, so I only have some aesthetical fixes.

this manual creation of the tree is such a pain, especially when some of the tests are trying to parse it. Need to create this using python instead.
